lang/cs/Org.Apache.REEF.Common/Telemetry/MetricTracker.cs

```diff
@@ -0,0 +1,232 @@
+        /// </summary>
+        [JsonProperty]
+        internal int ChangesSinceLastSink;
+
+        private IDisposable _unsubscriber;
+
+        /// <summary>
+        /// Constructor for MetricData called when metric is registered.
+        /// </summary>
+        /// <param name="metric"></param>
+        /// <param name="initialValue"></param>
+        internal MetricTracker(IMetric metric)
+        {
+            Subscribe(metric);
+            ChangesSinceLastSink = 0;
+            KeepUpdateHistory = metric.KeepUpdateHistory;
+            Records = KeepUpdateHistory ? new ConcurrentQueue<MetricRecord>() : null;
+            Records?.Enqueue(CreateMetricRecord(metric));
+        }
+
+        [JsonConstructor]
+        internal MetricTracker(IMetric metric, int changesSinceLastSink, ConcurrentQueue<MetricRecord> records, bool keepUpdateHistory)
+        {
+            Metric = metric;
+            Records = records;
+            KeepUpdateHistory = keepUpdateHistory;
+            ChangesSinceLastSink = changesSinceLastSink;
+        }
+
+        /// <summary>
+        /// Flush records currently held in the records queue.
+        /// </summary>
+        /// <returns>A queue containing all the flushed records.</returns>
+        internal IEnumerable<MetricRecord> FlushChangesSinceLastSink()
+        {
+            ConcurrentQueue<MetricRecord> records = new ConcurrentQueue<MetricRecord>();
+            if (Records != null)
+            {
+                while (Records.TryDequeue(out MetricRecord record))
+                {
+                    records.Enqueue(record);
+                }
+            }
+            else
+            {
+                // Records will be empty only on eval side when tracker doesn't keep history.
+                records.Enqueue(CreateMetricRecord(Metric));
+            }
+            Interlocked.Exchange(ref ChangesSinceLastSink, 0);
+            return records;
+        }
+
+        /// <summary>
+        /// When new metric data is received, update the value and records so it reflects the new data.
+        /// Called when Driver receives metrics from Evaluator.
+        /// </summary>
+        /// <param name="metric">Metric data received.</param>
+        internal MetricTracker UpdateMetric(MetricTracker metric)
+        {
+            if (metric.ChangesSinceLastSink > 0)
+            {
+                var recordsToAdd = metric.GetMetricRecords();
+                if (KeepUpdateHistory)
+                {
+                    foreach(MetricRecord record in recordsToAdd)
+                    {
+                        Records.Enqueue(record);
+                    }
+                }
+                else
+                {
+                    Interlocked.Exchange(ref Records, (ConcurrentQueue<MetricRecord>)recordsToAdd);
+                }
+            }
+            Interlocked.Add(ref ChangesSinceLastSink, metric.ChangesSinceLastSink);
```

Comment:
I took out the ChangesSinceLastSink variable to avoid having to use a lock (@singlis and I discussed this at length and wanted to avoid the performance hit.) Instead the sink threshold directly checks the number of cached records to decide whether to sink or not.